### PR TITLE
Updated ECMAScriptExpression refined regex

### PR DESCRIPTION
**Pull request by kshakir**

### CODE REVIEW

File: cwl/src/test/scala/wdl4s/cwl/CwlWorkflowWomSpec.scala [hunk 1]
```diff
@@ -124,10 +170,10 @@ class CwlWorkflowWomSpec extends FlatSpec with Matchers {
     val wc = nodes.collectFirst({ case wc: CallNode if wc.name == "wc" => wc }).get
 
     ps.upstream shouldBe empty
-    cgrep.upstream.filter(_ eq ps).size shouldBe 1
-    cgrep.upstream.filter(_ eq cgrepPatternInput).size shouldBe 1
+    cgrep.upstream.count(_ eq ps) shouldBe 1
+    cgrep.upstream.count(_ eq cgrepPatternInput) shouldBe 1
```
Comment (cjllanwarne) [inline cwl/src/test/scala/wdl4s/cwl/CwlWorkflowWomSpec.scala]: ToL: Maybe a new matcher than lets us easily do `Set[GraphNode]` reference equality would be useful here?

File: cwl/src/test/scala/wdl4s/cwl/CwlWorkflowWomSpec.scala [hunk 2]
```diff
@@ -124,10 +170,10 @@ class CwlWorkflowWomSpec extends FlatSpec with Matchers {
...
 
-    wc.upstream.filter(_ eq ps).size shouldBe 1
+    wc.upstream.count(_ eq ps) shouldBe 1
```
Comment (cjllanwarne) [inline cwl/src/test/scala/wdl4s/cwl/CwlWorkflowWomSpec.scala]: In the mean time, do we want to also check that `wc.upstream.size should be(1)`, eg?

File: cwl/src/test/scala/wdl4s/cwl/CwlWorkflowWomSpec.scala [hunk 3]
```diff
@@ -96,7 +97,52 @@ class CwlWorkflowWomSpec extends FlatSpec with Matchers {
     }
   }
 
-  "A WdlNamespace for 3step" should "provide conversion to WOM" in {
+  behavior of "A WdlNamespace for 3step"
```
Comment (cjllanwarne) [inline cwl/src/test/scala/wdl4s/cwl/CwlWorkflowWomSpec.scala]: `WdlNamespace` is wrong here

File: cwl/src/test/scala/wdl4s/cwl/CwlWorkflowWomSpec.scala [hunk 4]
```diff
@@ -96,7 +97,52 @@ class CwlWorkflowWomSpec extends FlatSpec with Matchers {
...
+
+  it should "parse ECMAScript expressions for arguments" in {
+    val wf = decodeAllCwl(rootPath / "three_step.cwl").map {
+      _.select[Workflow].get
+    }.value.unsafeRunSync.fold(error => throw new RuntimeException(s"broken parse! msg was $error"), identity)
+
+    // The second step (aka 1) step should be cgrep
+    val run: WorkflowStep.Run = wf.steps.apply(1).run
+    val commandLineTool: CommandLineTool = run.select[CommandLineTool].get
```
Comment (cjllanwarne) [inline cwl/src/test/scala/wdl4s/cwl/CwlWorkflowWomSpec.scala]: ToL: Even in tests, I'd prefer a nice `.getOrElse(fail("..."))` message to a generic `None.get exception`

File: cwl/src/test/scala/wdl4s/cwl/CwlWorkflowWomSpec.scala [hunk 5]
```diff
@@ -96,7 +97,52 @@ class CwlWorkflowWomSpec extends FlatSpec with Matchers {
...
+      (3, Inr(Inl("|"))),
+      (4, Inr(Inl("wc"))),
+      (5, Inr(Inl("-l")))
+    )
+
+    forAll(tests) { (index, result) =>
+      val argument: CommandLineTool.Argument = commandLineTool.arguments.get.apply(index)
+      val commandLineBinding: CommandLineBinding = argument.select[CommandLineBinding].get
+      val stringOrExpression: CommandLineTool.StringOrExpression = commandLineBinding.valueFrom.get
+      stringOrExpression should be(result)
```
Comment (cjllanwarne) [inline cwl/src/test/scala/wdl4s/cwl/CwlWorkflowWomSpec.scala]: ToL: any chance of making each case a `it should "correctly identify the expression X" in {}` case?

**Review by cjllanwarne — APPROVED** [review]
Nice! Consider the annoying "wouldn't it be nice" comments to be 100% optional

**Comment by Horneth** [comment]
Cool ! 👍 

[Approved with PullApprove](https://pullapprove.com/broadinstitute/wdl4s/pull-request/215/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=broadinstitute/wdl4s)

File: cwl/src/test/scala/wdl4s/cwl/CwlWorkflowWomSpec.scala (hunk 1 above)
Comment (kshakir) [reply cwl/src/test/scala/wdl4s/cwl/CwlWorkflowWomSpec.scala]: Switched away from reference equality, but is now using a scalatest Matcher.

File: cwl/src/test/scala/wdl4s/cwl/CwlWorkflowWomSpec.scala [hunk 6]
```diff
@@ -124,10 +184,8 @@ class CwlWorkflowWomSpec extends FlatSpec with Matchers {
     val wc = nodes.collectFirst({ case wc: CallNode if wc.name == "wc" => wc }).get
 
     ps.upstream shouldBe empty
-    cgrep.upstream.filter(_ eq ps).size shouldBe 1
-    cgrep.upstream.filter(_ eq cgrepPatternInput).size shouldBe 1
-
-    wc.upstream.filter(_ eq ps).size shouldBe 1
+    cgrep.upstream should contain theSameElementsAs Set(ps, cgrepPatternInput)
```
Comment (danbills) [inline cwl/src/test/scala/wdl4s/cwl/CwlWorkflowWomSpec.scala]: does `theSameElementsAs` do reference equality?  Important as that assertion broke this test earlier as we were creating duplicate equivalent instances.

File: cwl/src/test/scala/wdl4s/cwl/CwlWorkflowWomSpec.scala [hunk 7]
```diff
@@ -96,6 +98,64 @@ class CwlWorkflowWomSpec extends FlatSpec with Matchers {
...
 
+  behavior of "A decoded CWL 3step"
+
+  /**
+    * Runtime version of `refineMV[ECMAScriptExpression]("$(foo)")` as I cannot summon the dark magic for compile
+    * time checks.
+    *
+    * Compiler output after `import eu.timepit.refined._`:
+    *
+    * Error:(_, _) could not find implicit value for parameter v: eu.timepit.refined.api.Validate[String,wdl4s.cwl.ECMAScriptExpression]
```
Comment (danbills) [inline cwl/src/test/scala/wdl4s/cwl/CwlWorkflowWomSpec.scala]: is this perhaps how it tells you that it didn't match?

File: cwl/src/test/scala/wdl4s/cwl/CwlWorkflowWomSpec.scala [hunk 8]
```diff
@@ -96,6 +98,64 @@ class CwlWorkflowWomSpec extends FlatSpec with Matchers {
...
+    */
+  private def toECMAScriptExpression(unrefined: String): ECMAScriptExpression = {
+    eu.timepit.refined.api.RefType.applyRef[ECMAScriptExpression](unrefined).right
+      .getOrElse(fail(s"$unrefined was not an ECMAScriptExpression"))
+  }
+
+  private val stringOrExpressionTests = Table(
+    ("index", "result"),
+    (0, Inr(Inl("grep"))),
+    (1, Inl(toECMAScriptExpression("$(inputs.pattern)"))),
```
Comment (danbills) [inline cwl/src/test/scala/wdl4s/cwl/CwlWorkflowWomSpec.scala]: ToL these look like low-level declarations of Coproducts?  I think you can do `Coproduct[StringOrExpression]("grep")`.  Arguably easier to read/maintain.

File: cwl/src/test/resources/three_step.cwl [hunk 9]
```diff
@@ -54,7 +54,7 @@ steps:
     arguments:
     - valueFrom: grep
       shellQuote: false
-    - valueFrom: $(inputs.pattern).
```
Comment (danbills) [inline cwl/src/test/resources/three_step.cwl:57]: 🤦‍♂️

File: cwl/src/test/scala/wdl4s/cwl/CwlWorkflowWomSpec.scala [hunk 10]
```diff
@@ -96,6 +98,64 @@ class CwlWorkflowWomSpec extends FlatSpec with Matchers {
...
+
+    // The second step (aka 1) step should be cgrep
+    val run: WorkflowStep.Run = wf.steps.apply(1).run
+    val commandLineTool: CommandLineTool = run.select[CommandLineTool].getOrElse(fail(s"$run wasn't a CommandLineTool"))
+
+    commandLineTool.id.get should include("cgrep")
+    commandLineTool
+  }
+
+  forAll(stringOrExpressionTests) { (index, expected) =>
```
Comment (danbills) [inline cwl/src/test/scala/wdl4s/cwl/CwlWorkflowWomSpec.scala:140]: TIL `Table` and `forAll`.  Nice!

File: cwl/src/test/scala/wdl4s/cwl/CwlWorkflowWomSpec.scala (hunk 7 above)
Comment (kshakir) [reply cwl/src/test/scala/wdl4s/cwl/CwlWorkflowWomSpec.scala]: The same string matches at runtime. ☹️ Not sure what I'm doing wrong.

Comment (danbills) [reply cwl/src/test/scala/wdl4s/cwl/CwlWorkflowWomSpec.scala]: Ok I think I figured it out.  The `MatchesRegex` section of `ECMAScriptExpression` is the "predicate", so when you are declaring w/ `refineMV`, you use _only_ that.  

observe:
```scala
scala> val foo = refineMV[MatchesRegex[W.`"""\\$\\([^)]*\\)"""`.T]]("$(foo)")
foo: eu.timepit.refined.api.Refined[String,eu.timepit.refined.string.MatchesRegex[String("\\$\\([^)]*\\)")]] = $(foo)

scala> type Y = MatchesRegex[W.`"""\\$\\([^)]*\\)"""`.T]
defined type alias Y

scala> val foo = refineMV[Y]("$(foo)")
foo: eu.timepit.refined.api.Refined[String,Y] = $(foo)

scala> val x: ECMAScriptExpression = foo
x: wdl4s.cwl.ECMAScriptExpression = $(foo)
```

Comment (mcovarr) [reply cwl/src/test/scala/wdl4s/cwl/CwlWorkflowWomSpec.scala]: cool, and it looks like that type can be used at compile and runtime.

File: cwl/src/test/scala/wdl4s/cwl/CwlWorkflowWomSpec.scala [hunk 11]
```diff
@@ -96,6 +99,50 @@ class CwlWorkflowWomSpec extends FlatSpec with Matchers {
...
+    (2, Coproduct[CommandLineTool.StringOrExpression](refineMV[MatchesECMAScriptExpression]("$(inputs.file)"))),
+    (3, Coproduct[CommandLineTool.StringOrExpression]("|")),
+    (4, Coproduct[CommandLineTool.StringOrExpression]("wc")),
+    (5, Coproduct[CommandLineTool.StringOrExpression]("-l"))
+  )
+
+  private def getTestName(stringOrExpression: CommandLineTool.StringOrExpression): String = {
+    stringOrExpression match {
+      case Inl(expression) => s"expression ${expression.value}"
+      case Inr(Inl(string)) => s"string $string"
```
Comment (mcovarr) [inline cwl/src/test/scala/wdl4s/cwl/CwlWorkflowWomSpec.scala]: Is this the Right Way to type match with `Coproduct`s?

Comment (mcovarr) [reply cwl/src/test/scala/wdl4s/cwl/CwlWorkflowWomSpec.scala]: ToL a little unintuitive that `StringOrExpression` puts the expression type first if we're going to be doing ordered matches like this.

**Comment by cjllanwarne** [comment]
RE-👍

File: cwl/src/test/scala/wdl4s/cwl/CwlWorkflowWomSpec.scala (hunk 11 above)
Comment (danbills) [reply cwl/src/test/scala/wdl4s/cwl/CwlWorkflowWomSpec.scala]: A Poly1 & Fold is the Right Way ™️

**Review by mcovarr — APPROVED** [review]
Poly1 & Fold that one bit then :+1:

**Comment by kshakir** [comment]
Let me know if there's anything else. Once tests pass will be rebasing+squashing then merging.

File: cwl/src/test/scala/wdl4s/cwl/CwlWorkflowWomSpec.scala [hunk 12]
```diff
@@ -96,6 +99,56 @@ class CwlWorkflowWomSpec extends FlatSpec with Matchers {
...
+    (1, Coproduct[CommandLineTool.StringOrExpression](refineMV[MatchesECMAScriptExpression]("$(inputs.pattern)"))),
+    (2, Coproduct[CommandLineTool.StringOrExpression](refineMV[MatchesECMAScriptExpression]("$(inputs.file)"))),
+    (3, Coproduct[CommandLineTool.StringOrExpression]("|")),
+    (4, Coproduct[CommandLineTool.StringOrExpression]("wc")),
+    (5, Coproduct[CommandLineTool.StringOrExpression]("-l"))
+  )
+
+  private def getTestName(stringOrExpression: CommandLineTool.StringOrExpression): String = {
+    object StringOrExpressionToTestName extends Poly1 {
+      implicit def caseECMAScriptExpression: Case.Aux[ECMAScriptExpression, String] = {
```
Comment (mcovarr) [inline cwl/src/test/scala/wdl4s/cwl/CwlWorkflowWomSpec.scala:115]: Huh, I thought the ascriptions were way worse than this and that was why we weren't including them on `Poly1`s, but this doesn't seem bad to me at all.

**Comment by mcovarr** [comment]
looks good, superfluous :+1:

File: cwl/src/test/scala/wdl4s/cwl/CwlWorkflowWomSpec.scala (hunk 12 above)
Comment (kshakir) [reply cwl/src/test/scala/wdl4s/cwl/CwlWorkflowWomSpec.scala:115]: The default IntellJ ascriptions don't know about Shapeless' [`Aux`-pattern](http://www.vlachjosef.com/aux-pattern-evolution/) and end up explicitly defining the inner `type`. They're all easy enough to type by hand.

Comment (mcovarr) [reply cwl/src/test/scala/wdl4s/cwl/CwlWorkflowWomSpec.scala:115]: Very nice, I'd like to continue adding these then.

**Comment by kshakir** [comment]
No idea why coveralls isn't refreshing. Merging

**Pull request merged**